blog-content/the-complete-guide-to-opentelemetry-in-next-js.md

```diff
@@ -0,0 +1,394 @@
+}
+```
+
+You can observe spans for:
+✅ Incoming request (GET /api/user)
+✅ Response timing
+✅ Any internal or external API calls
+
+#### Tracing Page Router (Pages Directory)
+
+For traditional Next.js pages (pages/index.js):
+
+```javascript
+export default function HomePage() {
+  return <h1>Welcome to Next.js</h1>;
+}
+```
+
+When a user navigates to /, Next.js automatically creates spans for:
+- next.js / (page load)
+- next.js render /index
+- fetch spans (if the page makes API calls)
+
+#### Tracing App Router (App Directory)
+
+If you’re using the Next.js App Router (app/page.tsx), spans will include:
+- Page load spans (next.js /app/page.tsx)
+- Server component execution spans (next.js /layout.tsx)
+- Data fetching spans for fetch() calls in server components
+
+For example, an app router component:
+
+```javascript
+export default function Page() {
+  return <h1>Welcome to the App Router!</h1>;
+}
+```
+will automatically generate spans for rendering and hydration.
+
+### Logging in OpenTelemetry
+
+To correlate logs with traces, install OpenTelemetry logging:
+
+
+```bash
+yarn add @opentelemetry/sdk-logs
+```
+
+Modify `otel.js`:
+
+```javascript
+import { LoggerProvider } from '@opentelemetry/sdk-logs';
+import { ConsoleLogger } from '@opentelemetry/sdk-logs';
+
+const loggerProvider = new LoggerProvider();
+const logger = loggerProvider.getLogger('nextjs-logger');
+
+logger.emit({
+  severityText: 'INFO',
+  body: 'Application started',
+});
+```
+This allows you to log messages that correlate with traces.
+
+### Capturing Exceptions with Spans
+
+To capture exceptions inside traces, you can manually add error attributes to spans. Modify your API route:
+
+```javascript
+import { trace } from '@opentelemetry/api';
+
+export default async function handler(req, res) {
+  const span = trace.getActiveSpan();
+  
+  try {
+    throw new Error('Something went wrong!');
+  } catch (error) {
+    if (span) {
+      span.setAttribute('error', true);
+      span.recordException(error);
+    }
+    res.status(500).json({ error: 'Internal Server Error' });
+  }
+}
+```
+This ensures that the error is captured within the OpenTelemetry trace and can be visualized in your tracing backend.
+
+### Exporting Metrics
+
+Next.js applications often benefit from metrics like request count, latency, and errors. Modify `otel.js`:
+
+```javascript
+import { MeterProvider } from '@opentelemetry/sdk-metrics';
+
+const meter = new MeterProvider().getMeter('nextjs-meter');
+const requestCounter = meter.createCounter('http_requests_total', {
+  description: 'Counts total HTTP requests',
+});
+
+export function trackRequest() {
+  requestCounter.add(1);
+}
+```
+
+Then, use it in an API route:
+```javascript
+import { trackRequest } from '../../otel';
+
+export default function handler(req, res) {
+  trackRequest();
+  res.status(200).json({ message: 'Metrics tracked!' });
+}
+```
+
+### Simplifying with @vercel/otel
```

Comment:
WDYT about calling out some of the complexities + limitations of the manual instrumentation we've built up so far before going into @vercel/otel?

Reply:
reworked this!